File: pipeline/reimpl/cbcus01c.py
"""
Reimplementation of CBCUS01C — CardDemo Batch Customer File Reader.

Function: Read the customer VSAM KSDS sequentially and display each record.
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Callable

from .carddemo_data import CustomerRecord


@dataclass
class ProcessResult:
    records_read: int = 0
    display_lines: list[str] = field(default_factory=list)


def process_customer_file(
    customers: list[CustomerRecord],
    logger: Callable[[str], None] = print,
) -> ProcessResult:
    """Process all customer records — mirrors CBCUS01C PROCEDURE DIVISION."""
    result = ProcessResult()
    logger("START OF EXECUTION OF PROGRAM CBCUS01C")

    for cust in customers:
        result.records_read += 1
        line = _format_customer_record(cust)
        logger(line)
        result.display_lines.append(line)

    logger("END OF EXECUTION OF PROGRAM CBCUS01C")
    return result


def _format_customer_record(cust: CustomerRecord) -> str:
    """DISPLAY CUSTOMER-RECORD — all fields formatted."""
    return (
        f"CUST-ID:{cust.cust_id:09d} "
        f"NAME:{cust.cust_first_name.strip()} {cust.cust_middle_name.strip()} "
        f"{cust.cust_last_name.strip()} "
        f"SSN:{cust.cust_ssn:09d} "
        f"DOB:{cust.cust_dob_yyyy_mm_dd} "
        f"FICO:{cust.cust_fico_credit_score:03d} "
        f"ZIP:{cust.cust_addr_zip}"
    )

File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/Cbcus01c.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.runner.ProgramRunner;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Java reimplementation of CardDemo batch program CBCUS01C — Customer File Reader.
 *
 * <p>Mirrors {@code pipeline/reimpl/cbcus01c.py} byte-for-byte. Python is the
 * source of truth.
 */
public class Cbcus01c implements ProgramRunner {

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        // Seed data — 2 customer records
        // Fields: custId, firstName, middleName, lastName, ssn, dob, fico, zip
        Object[][] customers = {
            {1, "John", "A", "Doe",   123456789, "1990-01-15", 750, "10001"},
            {2, "Jane", "B", "Smith", 987654321, "1985-07-22", 680, "90210"},
        };

        Map<String, String> out = new LinkedHashMap<>();
        out.put("RECORDS_READ", String.valueOf(customers.length));

        for (int i = 0; i < customers.length; i++) {
            Object[] c = customers[i];
            String line = String.format(
                "CUST-ID:%09d NAME:%s %s %s SSN:%09d DOB:%s FICO:%03d ZIP:%s",
                ((Number) c[0]).intValue(),
                ((String) c[1]).trim(),
                ((String) c[2]).trim(),
                ((String) c[3]).trim(),
                ((Number) c[4]).intValue(),
                (String) c[5],
                ((Number) c[6]).intValue(),
                (String) c[7]
            );
            out.put("DISPLAY_" + i, line);
        }

        return out;
    }
}
